README.md
# TinyMCE Plugin #

## Description ##
This plugin provides TinyMCE 4 for editing messages and templates within phplist.

It also integrates the elFinder file manager to provide file upload and selection.
## Compatibility ###

TinyMCE and elFinder are compatible with all the major browsers, see the TinyMCE site <http://www.tinymce.com/index.php>
and the elFinder site <https://github.com/Studio-42/elFinder/>

## Dependencies ##

This plugin is for phplist 3.

Requires php version 5.3 or later.

## Installation ##

For guidance on installing and configuring the plugin see the plugin's page within the phplist documentation site <https://resources.phplist.com/plugin/tinymce>.

## Donation ##

This plugin is free but if you install and find it useful then a donation to support further development is greatly appreciated.

[![Donate](https://www.paypalobjects.com/en_US/i/btn/btn_donate_LG.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=W5GLX53WDM7T4)

## Version history ##

    version     Description
    2.4.2+20231202  Always edit templates as full page
    2.4.1+20231202  Allow upload dir per admin not to apply to super users
    2.4.0+20231128  Allow upload subdirectory for each admin
    2.3.0+20220903  Allow tinymce window width not to be set so that it fills the surrounding element.
    2.2.0+20210129  Use the local copy of tinymce instead of the cloud version
    2.1.0+20161018  Load tinymce from its CDN
    2.0.2+20160615  GitHub #3, do not specify protocol for jquery files
    2.0.1+20150914  Allow full page editing
    2015-07-15      GitHub #1, use https for jquery files
    2014-07-13      Accumulated minor changes
    2014-03-24      Include host name in links
    2013-11-05      Initial version for phplist 3


changelog.txt
version 2.4.2+20231202
      Always edit templates as full page

version 2.4.1+20231202
      Allow image dir per admin not to apply to super users

version 2.4.0+20231128
      Revise CS Fixer and gitignore
      Allow upload subdirectory for each admin
      Fix php8 errors and deprecations

version 2.3.0+20220903
      Allow tinymce window width not to be set so that it fills the surrounding element.

version 2.2.0+20210129
      Add local copy of tinymce 4.9.11
      Use the local copy of tinymce instead of the cloud version

version 2.1.0+20161018
      Load tinymce from its CDN
      Add item to tinymce menu to show its version
      coding standards changes
      Move most of the documentation to the phplist plugins page

version 2.0.2+20160615
Arnaud Virlet (1):
     GitHub #3, update url for ajax.googleapis.com (avoid blocked mixed active content error in browser)

version 2.0.1+20150914
      Allow full page editing
      Don't escape & in javascript

version 2015-07-15
      GitHub #1, use https instead of http for jquery files

version 2015-05-10
      Add dependency checks

version 2014-07-13
      Ensure host name is included in links
      Display warning if image directory is not writeable
      Correct unescaped &

version 2014-03-24
      Internal changes
      Ensure host name is included in links
      Allow UPLOADIMAGES_DIR to have leading and trailing /

version 2013-11-05
      Initial release


plugins/TinyMCEPlugin.php
<?php
/**
 * TinyMCEPlugin for phplist.
 *
 * This file is a part of TinyMCEPlugin.
 *
 * This plugin is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * This plugin is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * @category  phplist
 *
 * @author    Duncan Cameron
 * @copyright 2013 Duncan Cameron
 * @license   http://www.gnu.org/licenses/gpl.html GNU General Public License, Version 3
 */
class TinyMCEPlugin extends phplistPlugin
{
    const VERSION_FILE = 'version.txt';
    const CODE_DIR = '/TinyMCEPlugin/';
    /*
     *  Private variables
     */
    private $elEnabled;

    /*
     *  Inherited variables
     */
    public $name = 'Tiny MCE plugin';
    public $editorProvider = true;
    public $authors = 'Duncan Cameron';
    public $description = 'Provides the TinyMCE editor for editing messages and templates.';
    public $enabled = 1;

    private function elFinderScript($function)
    {
        $elUrl = './?pi=TinyMCEPlugin&page=elfinder_standalone';
        $html = <<<END
<script type='text/javascript'>
$function = function(callback) {
    window.elFinder = {};
    window.elFinder.callBack = function(url) {
        callback(url);
        window.elFinder = null;
    };
    window.open('$elUrl', '', 'width=900,height=450');
};
</script>
END;

        return $html;
    }

    private function editorScript($fieldname, $width, $height, $toolbar)
    {
        $html = '';
        $tinymceUrl = getConfig('tinymce_url');

        // If the tinymce URL is for version 4 then change it to the default
        if ($tinymceUrl == '//cdn.tinymce.com/4/tinymce.min.js') {
            $tinymceUrl = $this->settings['tinymce_url']['value'];
            saveConfig('tinymce_url', $tinymceUrl);
        }

        if (substr($tinymceUrl, -15) != '/tinymce.min.js') {
            $html .= sprintf(
                '<div class="note error">tinyMCE is not available because the setting for the URL of tinymce.min.js "%s" is incorrect.</div>',
                $tinymceUrl
            );
        }

        if (!is_writeable($dir = $_SERVER['DOCUMENT_ROOT'] . '/' . trim(UPLOADIMAGES_DIR, '/'))) {
            $html .= sprintf(
                '<div class="note error">The image upload directory "%s" does not exist or is not writeable.</div>',
                htmlspecialchars($dir)
            );
        }
        $settings = array();

        $fullPageSetting = '';

        if ($fieldname == 'template' || ($fieldname == 'message' && getConfig('tinymce_fullmessage'))) {
            $fullPageSetting = 'settings.plugins.push("fullpage");';
        }

        $config = getConfig('tinymce_config');

        if ($config) {
            $settings[] = trim(trim($config), ',');
        }
        if ($width) {
            $settings[] = "width: $width";
        }

        if ($height) {
            $settings[] = "height: $height";
        }

        if ($toolbar) {
            $toolbar = trim($toolbar, '"');
            $settings[] = "toolbar: \"$toolbar\"";
        }
        $configSettings = implode(",\n", $settings);

        $html .= <<<END
<style>
    body.send .panel .content .mce-tinymce div {margin: 0px}
</style>
<script src="$tinymceUrl"></script>
<script type="text/javascript">
settings = {
    setup: function(ed) {
        ed.addMenuItem('info', {
            'text' : 'Info',
            'context' : 'tools',
            'icon' : 'help',
            'onclick' : function() {
                tinymce.activeEditor.windowManager.alert('TinyMCE ' + tinymce.majorVersion + '.' + tinymce.minorVersion);
            }
        });
    },
    selector: "textarea.editable",
    file_browser_callback: elFinderBrowser,
    relative_urls: false,
    remove_script_host: false,
    $configSettings
};
$fullPageSetting
tinymce.init(settings);
function elFinderBrowser (field_name, url, type, win) {
  tinymce.activeEditor.windowManager.open({
    file: './?pi=TinyMCEPlugin&page=elfinder_tinymce',
    title: 'elFinder',
    width: 900,
    height: 450,
    resizable: 'yes'
  }, {
    setUrl: function (url) {
      win.document.getElementById(field_name).value = url;
    }
  });
  return false;
}
</script>
END;

        return $html;
    }

    public function __construct()
    {
        $this->elEnabled = defined('UPLOADIMAGES_DIR') && UPLOADIMAGES_DIR !== false;
        $this->coderoot = dirname(__FILE__) . self::CODE_DIR;
        $this->version = (is_file($f = $this->coderoot . self::VERSION_FILE))
            ? file_get_contents($f)
            : '';
        $tinymceUrl = substr(PLUGIN_ROOTDIR, 0, 1) == '/' ? PLUGIN_ROOTDIR : $GLOBALS['pageroot'] . '/admin/' . PLUGIN_ROOTDIR;
        $tinymceUrl .= self::CODE_DIR . 'tinymce/js/tinymce/tinymce.min.js';
        $elPath = substr(PLUGIN_ROOTDIR, 0, 1) == '/' ? PLUGIN_ROOTDIR : $GLOBALS['pageroot'] . '/admin/' . PLUGIN_ROOTDIR;
        $elPath .= self::CODE_DIR . 'elfinder';

        $this->settings = array(
            'tinymce_url' => array(
              'value' => $tinymceUrl,
              'description' => 'URL of tinymce.min.js',
              'type' => 'text',
              'allowempty' => 0,
              'category' => 'TinyMCE',
            ),
            'tinymce_config' => array(
              'value' => '',
              'description' => 'Custom configuration settings',
              'type' => 'textarea',
              'allowempty' => 1,
              'category' => 'TinyMCE',
            ),
            'tinymce_width' => array(
              'value' => 0,
              'description' => 'Width in px of TinyMCE window. Enter 0 to not set the width',
              'type' => 'integer',
              'allowempty' => true,
              'min' => 0,
              'max' => 800,
              'category' => 'TinyMCE',
            ),
            'tinymce_height' => array(
              'value' => 600,
              'description' => 'Height in px of TinyMCE window',
              'type' => 'integer',
              'allowempty' => 0,
              'min' => 100,
              'max' => 800,
              'category' => 'TinyMCE',
            ),
            'tinymce_fullmessage' => array(
              'description' => 'Allow messages to be edited as full HTML pages',
              'type' => 'boolean',
              'value' => false,
              'allowempty' => true,
              'category' => 'TinyMCE',
            ),
        );
        $this->settings['tinymce_config']['value'] = <<<END
toolbar: "insertfile undo redo | styleselect | bold italic | alignleft aligncenter alignright alignjustify | bullist numlist | link image",
plugins: [
        "advlist autolink lists link image charmap print preview anchor",
        "searchreplace visualblocks code fullscreen",
        "insertdatetime media table contextmenu paste"
    ]
END;

        if ($this->elEnabled) {
            $this->settings += array(
                'elfinder_path' => array(
                  'value' => $elPath,
                  'description' => 'path to elFinder',
                  'type' => 'text',
                  'allowempty' => 0,
                  'category' => 'TinyMCE',
                ),
            );
        }
        parent::__construct();
    }

    public function dependencyCheck()
    {
        global $editorplugin;

        return array(
            'PHP version 5.3.0 or greater' => version_compare(PHP_VERSION, '5.3') > 0,
            'No other editor enabled' => empty($editorplugin) || $editorplugin == 'TinyMCEPlugin',
        );
    }

    public function adminmenu()
    {
        return array();
    }

    public function editor($fieldname, $content)
    {
        $width = getConfig('tinymce_width');
        $height = getConfig('tinymce_height');

        return $this->createEditor($fieldname, $content, $width, $height);
    }

    public function createEditor($fieldname, $content, $width = null, $height = null, $toolbar = null)
    {
        $fieldname = htmlspecialchars($fieldname);
        $content = htmlspecialchars($content);
        $html = $this->editorScript($fieldname, $width, $height, $toolbar) . <<<END
<textarea class="editable" name="$fieldname">$content</textarea>
END;

        return $html;
    }

    public function createImageBrowser($function)
    {
        static $firstTime = true;

        if ($firstTime) {
            $firstTime = false;
            $html = $this->elFinderScript($function);
        } else {
            $html = '';
        }

        return $html;
    }
}
